README.md
# Devopsbot - Development Operations Bot
Devopsbot is a Slack bot written in Go using the [Slack API in Go](https://github.com/slack-go/slack).
It improves development efficiency by automating tasks such as:
- Declaring incidents
- Resolving incidents

The bot essentially automates the Incident Command System (ICS).

## Design Principles
The application is built around three key principles:
1. Everything the bot can do, any person is able to do - if it goes down and is unavailable, it will not block anyone
1. Secrets are maintained externally, for example via Hashicorp Vault
1. Configuration is maintained externally, for example via Kubernetes config maps

## Documentation

For more information, see the [Devopsbot documentation](https://karl-johan-grahn.github.io/devopsbot/).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.15.21] - 2022-07-19
### Update
- Update dependency docker.io/library/alpine to v3.16.1

## [0.15.20] - 2022-07-13
### Update
- Update dependency docker.io/library/golang to v1.18.4

## [0.15.19] - 2022-06-29
### Update
- Update module github.com/stretchr/testify to v1.8.0

## [0.15.18] - 2022-06-26
### Update
- Update module github.com/stretchr/testify to v1.7.5

## [0.15.17] - 2022-06-21
### Update
- Update module github.com/stretchr/testify to v1.7.4

## [0.15.16] - 2022-06-21
### Update
- Update module github.com/spf13/cobra to v1.5.0

## [0.15.15] - 2022-06-20
### Update
- Update module github.com/stretchr/testify to v1.7.3

## [0.15.14] - 2022-06-13
### Update
- Update errata-ai/vale-action action to v2

## [0.15.13] - 2022-06-13
### Update
- Update dependency docker.io/library/golang to v1.18.3

## [0.15.12] - 2022-06-13
### Update
- Update golang.org/x/term digest to 065cf7b

## [0.15.11] - 2022-06-13
### Update
- Update docker/build-push-action action to v3

## [0.15.10] - 2022-06-13
### Update
- Update docker/login-action action to v2

## [0.15.9] - 2022-06-13
### Update
- Update module github.com/spf13/viper to v1.12.0

## [0.15.8] - 2022-06-13
### Update
- Update module github.com/prometheus/client_golang to v1.12.2

## [0.15.7] - 2022-06-13
### Update
- Update module github.com/stretchr/testify to v1.7.2

## [0.15.6] - 2022-06-13
### Update
- Update dependency docker.io/library/alpine to v3.16.0

## [0.15.5] - 2022-06-13
### Update
- Update module github.com/spf13/cobra to v1.4.0

## [0.15.4] - 2022-06-13
### Update
- Update module github.com/rs/zerolog to v1.27.0

## [0.15.3] - 2022-06-13
### Update
- Update module github.com/nicksnyder/go-i18n/v2 to v2.2.0

## [0.15.2] - 2022-06-13
### Update
- Update docker/metadata-action action to v4

## [0.15.1] - 2022-06-13
### Update
- Update github/codeql-action action to v2

## [0.15.0] - 2022-06-13
### Update
- Update module github.com/slack-go/slack to v0.11.0

## [0.14.14] - 2022-05-13
### Update
- Update incident documentation

## [0.14.13] - 2022-04-01
### Fix
- Fix spelling

## [0.14.12] - 2022-04-01
### Updates
- Update module github.com/stretchr/testify to v1.7.1

## [0.14.11] - 2022-04-01
### Updates
- Update hadolint/hadolint-action action to v2

## [0.14.10] - 2022-04-01
### Updates
- Update actions/setup-go action to v3

## [0.14.9] - 2022-04-01
### Updates
- Update docker/login-action action to v1.14.1

## [0.14.8] - 2022-04-01
### Updates
- Update errata-ai/vale-action action to v1.5.0

## [0.14.7] - 2022-04-01
### Updates
- Update actions/checkout action to v3

## [0.14.6] - 2022-04-01
### Updates
- Update docker/build-push-action action to v2.10.0

## [0.14.5] - 2022-04-01
### Updates
- Update hadolint/hadolint-action action to v1.7.0

## [0.14.4] - 2022-02-13
### Updates
- Update docker/build-push-action action to v2.9.0

## [0.14.3] - 2022-02-13
### Updates
- Update module github.com/prometheus/client_golang to v1.12.1

## [0.14.2] - 2022-02-13
### Updates
- Update docker.io/library/golang Docker tag to v1.17.7

## [0.14.1] - 2022-02-13
### Updates
- Update module github.com/slack-go/slack to v0.10.2

## [0.14.0] - 2022-01-12
### Adds
- Check if bot is part of broadcast channel and that broadcast channel is not archived
- Do logging via UI to improve usability
- Update docker.io/library/golang Docker tag to v1.17.6
- Update github.com/prometheus/client_golang to v1.11.0

## [0.13.9] - 2021-12-23
### Updates
- Update golang.org/x/term commit hash to 03fcf44

## [0.13.8] - 2021-12-23
### Updates
- Update docker.io/library/alpine Docker tag to v3.15.0

## [0.13.7] - 2021-12-23
### Updates
- Update docker.io/library/golang Docker tag to v1.17.5

## [0.13.6] - 2021-12-23
### Updates
- Update module github.com/spf13/viper to v1.10.1

## [0.13.5] - 2021-12-23
### Updates
- Update module github.com/spf13/cobra to v1.3.0

## [0.13.4] - 2021-12-23
### Updates
- Update module github.com/rs/zerolog to v1.26.1

## [0.13.3] - 2021-12-23
### Updates
- Update errata-ai/vale-action action to v1.4.3

## [0.13.2] - 2021-12-23
### Updates
- Update docker/login-action action to v1.12.0

## [0.13.1] - 2021-12-23
### Updates
- Update docker/metadata-action action to v3.6.2

## [0.13.0] - 2021-12-22
### Updates
- When generating version file, enable matching non-annotated tags
- Slack does not yet allow users to create reminders recurring more often than once a day, so just create one that runs daily 30 min after the incident has been declared
- Include year in Slack channel name to decrease chance of having name creation conflicts and to make the name more explicit
- Describe incidents according to severity and impact

## [0.12.0] - 2021-11-29
### Adds
- Add broadcast channel as input option, add dispatch action when characters are entered, update to Slack Go API v0.10.0

## [0.11.0] - 2021-11-16
### Updates
- Update to Go 1.17.3

## [0.10.0] - 2021-11-03
### Updates
- Update UI such as update wording, add hints for responder and commander

## [0.9.0] - 2021-11-02
### Updates
- Update date handling

## [0.8.0] - 2021-10-27
### Updates
- French string updates and package updates

## [0.7.0] - 2021-10-25
### Updates
- Update strings and remove invitation of bot

## [0.6.0] - 2021-10-22
### Updates

bot/interactive.go
package bot

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/karl-johan-grahn/devopsbot/internal/middleware"
	"github.com/karl-johan-grahn/devopsbot/internal/wrappedcontext"
	"github.com/rs/zerolog"
	"github.com/slack-go/slack"
)

var channelNameRegex = regexp.MustCompile(`^[a-z0-9\-\_]+$`)
var incChannelNameRegex = regexp.MustCompile(`^inc_`)

const (
	alreadyInChannel   = "already_in_channel"
	valIncChName       = "\"%s\" - channel name must be non-empty, and contain only lowercase letters, numbers, hyphens, and underscores"
	valChosenIncChName = "#%s does not seem to be an incident channel"
)

// handleInteractive - a general handler for the /interactive endpoint
func (h *botHandler) handleInteractive(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	log := zerolog.Ctx(ctx)

	payload := &slack.InteractionCallback{}
	err := json.Unmarshal([]byte(r.FormValue("payload")), payload)
	if err != nil {
		err = middleware.NewHTTPError(fmt.Errorf("failed to parse interactive payload: %w", err), r)
		log.Error().Err(err).Interface("payload", payload).Send()
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	switch payload.Type {
	case slack.InteractionTypeBlockActions:
		if len(payload.ActionCallback.BlockActions) == 0 {
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		action := payload.ActionCallback.BlockActions[0]
		switch action.BlockID {
		case "incident_name":
			// This is triggered when pressing enter in the text input box
			if err := validateIncidentChannelName("incident_name", action.Value); err != nil {
				if uerr := h.updateView(ctx, payload, "incident_name", "declare_incident", fmt.Sprintf(valIncChName, action.Value), w); uerr != nil {
					uerr = middleware.NewHTTPError(uerr, r)
					log.Error().Err(uerr).Msg("updateView failed")
					w.WriteHeader(http.StatusInternalServerError)
					return
				}
			} else {
				if uerr := h.updateView(ctx, payload, "incident_name", "declare_incident",
					fmt.Sprintf("This will create this channel name: #%s", createChannelName(action.Value)), w); uerr != nil {
					uerr = middleware.NewHTTPError(uerr, r)
					log.Error().Err(uerr).Msg("updateView failed")
					w.WriteHeader(http.StatusInternalServerError)
					return
				}
			}
		case "incident_channel":
			channelID := action.SelectedConversation
			channel, _ := h.slackClient.GetConversationInfoContext(ctx, channelID, false)
			if err := validateChosenIncidentChannelName("incident_channel", channel.Name); err != nil {
				if uerr := h.updateView(ctx, payload, "incident_channel", "resolve_incident", fmt.Sprintf(valChosenIncChName, channel.Name), w); uerr != nil {
					uerr = middleware.NewHTTPError(uerr, r)
					log.Error().Err(uerr).Msg("updateView failed")
					w.WriteHeader(http.StatusInternalServerError)
					return
				}
			}
		default:
			log.Error().Str("BlockID", action.BlockID).Msg("unknown BlockID")
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		w.WriteHeader(http.StatusOK)
	case slack.InteractionTypeViewSubmission:
		callbackID := payload.View.CallbackID
		if callbackID == "" {
			log.Error().Msg("callbackID empty")
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		switch callbackID {
		case "declare_incident":
			if err := h.declareIncident(ctx, payload, w); err != nil {
				err = middleware.NewHTTPError(err, r)
				log.Error().Err(err).Msg("declareIncident failed")
				w.WriteHeader(http.StatusInternalServerError)
				return
			}
		case "resolve_incident":
			if err := h.resolveIncident(ctx, payload, w); err != nil {
				err = middleware.NewHTTPError(err, r)
				log.Error().Err(err).Msg("resolveIncident failed")
				w.WriteHeader(http.StatusInternalServerError)
				return
			}
		default:
			log.Error().Str("callbackID", callbackID).Msg("unknown callbackID")
			w.WriteHeader(http.StatusBadRequest)
			return
		}
		w.WriteHeader(http.StatusOK)
	default:
		log.Warn().Str("payload_type", string(payload.Type)).Msg("unknown interactive payload type")
		w.WriteHeader(http.StatusNotFound)
	}
}

func (h *botHandler) updateView(ctx context.Context, payload *slack.InteractionCallback, blockID string, callbackID string, updatedText string, w http.ResponseWriter) error {
	log := zerolog.Ctx(ctx)
	for _, b := range payload.View.Blocks.BlockSet {
		if input, ok := b.(*slack.InputBlock); ok && input.BlockID == blockID {
			input.Hint = slack.NewTextBlockObject(slack.PlainTextType, updatedText, false, false)
			input.DispatchAction = true
		}
	}
	mvr := slack.ModalViewRequest{
		Type:       payload.View.Type,
		Title:      payload.View.Title,
		Close:      payload.View.Close,
		Submit:     payload.View.Submit,
		Blocks:     payload.View.Blocks,
		CallbackID: callbackID,
	}
	_, err := h.slackClient.UpdateViewContext(ctx, mvr, payload.View.ExternalID, payload.View.Hash, payload.View.ID)
	if err != nil {
		log.Error().Err(err).Msg("Error updating view")
		w.WriteHeader(http.StatusInternalServerError)
		return err
	}
	return nil
}

type inputParams struct {
	incidentChannelName          string
	incidentSecurityRelated      bool
	incidentResponder            string
	incidentCommander            string
	incidentInvitees             []string
	incidentEnvironmentsAffected string
	incidentRegionsAffected      string
	IncidentSeverityLevel        string
	IncidentImpactLevel          string
	incidentSummary              string
	incidentDeclarer             string
	broadcastChannel             string
}

type validationError struct {
	errors map[string]string
}

// Error - conform to the error interface
func (e *validationError) Error() string {
	badFields := make([]string, len(e.errors))
	i := 0
	for _, k := range e.errors {
		badFields[i] = k
		i++
	}
	return fmt.Sprintf("validation errors with field(s): %s", strings.Join(badFields, ", "))
}

// validatePayload - validate incident payload
func validatePayload(ctx context.Context, payload *slack.InteractionCallback) error {
	incidentChannelName := createChannelName(payload.View.State.Values["incident_name"]["incident_name"].Value)
	return validateIncidentChannelName("incident_name", incidentChannelName)
}

// declareIncident - general handler for incident commands
func (h *botHandler) declareIncident(ctx context.Context, payload *slack.InteractionCallback, w http.ResponseWriter) error {
	if err := validatePayload(ctx, payload); err != nil {
		var verr *validationError
		if errors.As(err, &verr) {
			return postErrorResponse(ctx, verr.errors, w)
		}
		return err
	}
	incidentChannelName := createChannelName(payload.View.State.Values["incident_name"]["incident_name"].Value)
	incidentEnvironmentsAffected := make([]string, len(payload.View.State.Values["incident_environment_affected"]["incident_environment_affected"].SelectedOptions))
	for i, e := range payload.View.State.Values["incident_environment_affected"]["incident_environment_affected"].SelectedOptions {
		incidentEnvironmentsAffected[i] = e.Value
	}
	incidentRegionsAffected := make([]string, len(payload.View.State.Values["incident_region_affected"]["incident_region_affected"].SelectedOptions))
	for i, r := range payload.View.State.Values["incident_region_affected"]["incident_region_affected"].SelectedOptions {
		incidentRegionsAffected[i] = r.Value
	}
	incidentSecurityRelated := false
	if len(payload.View.State.Values["security_incident"]["security_incident"].SelectedOptions) > 0 {
		incidentSecurityRelated = payload.View.State.Values["security_incident"]["security_incident"].SelectedOptions[0].Value == "yes"
	}
	inputParams := &inputParams{
		broadcastChannel:             payload.View.State.Values["broadcast_channel"]["broadcast_channel"].SelectedOption.Value,
		incidentChannelName:          incidentChannelName,
		incidentSecurityRelated:      incidentSecurityRelated,
		incidentResponder:            payload.View.State.Values["incident_responder"]["incident_responder"].SelectedUser,
		incidentCommander:            payload.View.State.Values["incident_commander"]["incident_commander"].SelectedUser,
		incidentInvitees:             payload.View.State.Values["incident_invitees"]["incident_invitees"].SelectedUsers,
		incidentEnvironmentsAffected: strings.Join(incidentEnvironmentsAffected, ", "),
		incidentRegionsAffected:      strings.Join(incidentRegionsAffected, ", "),
		IncidentSeverityLevel:        payload.View.State.Values["incident_severity_level"]["incident_severity_level"].SelectedOption.Value,
		IncidentImpactLevel:          payload.View.State.Values["incident_impact_level"]["incident_impact_level"].SelectedOption.Value,
		incidentSummary:              payload.View.State.Values["incident_summary"]["incident_summary"].Value,
		incidentDeclarer:             payload.User.ID,
	}
	// Add incident responder and incident commander to the people to be invited to the incident channel
	inputParams.incidentInvitees = append(inputParams.incidentInvitees, inputParams.incidentResponder, inputParams.incidentCommander)

	// Create channel - should be done here because it will update the modal if there are errors
	incidentChannel, err := h.slackClient.CreateConversationContext(ctx, incidentChannelName, inputParams.incidentSecurityRelated)
	if err != nil {
		errorMessage := createUserFriendlyConversationError(err)
		return postErrorResponse(ctx, map[string]string{
			"incident_name": fmt.Sprintf("%s: <#%s>", errorMessage, incidentChannelName),
		}, w)
	}

	w.WriteHeader(http.StatusAccepted)

	// Do the rest via goroutine
	ctx = wrappedcontext.WrapContextValues(context.Background(), ctx)
	go h.doIncidentTasks(ctx, inputParams, incidentChannel)

	return nil
}

// doIncidentTasks - do incident creation tasks asynchronously
func (h *botHandler) doIncidentTasks(ctx context.Context, params *inputParams, incidentChannel *slack.Channel) {
	log := zerolog.Ctx(ctx)
	const sendError = "Could not send failure message"
	var securityMessage string
	if params.incidentSecurityRelated {
		securityMessage = "This is a security related incident - available by invitation only"
	} else {
		securityMessage = ""
	}
	// Set channel purpose and topic - they can be maximum 250 characters
	overview := fmt.Sprintf("*Environment affected:* %s\n"+
		"*Region affected:* %s\n"+
		"*Severity:* %s\n"+
		"*Impact:* %s\n"+
		"*Responder:* <@%s>\n"+
		"*Commander:* <@%s>\n"+
		"*Broadcast channel:* <#%s>\n\n"+
		"Declared by: <@%s>\n"+
		securityMessage,
		params.incidentEnvironmentsAffected, params.incidentRegionsAffected,
		params.IncidentSeverityLevel, params.IncidentImpactLevel,
		params.incidentResponder, params.incidentCommander, params.broadcastChannel, params.incidentDeclarer)
	if _, err := h.slackClient.SetPurposeOfConversationContext(ctx, incidentChannel.ID, overview); err != nil {
		if sendErr := h.sendMessage(ctx, params.broadcastChannel, slack.MsgOptionPostEphemeral(params.incidentDeclarer),
			slack.MsgOptionText(fmt.Sprintf("Failed to set purpose for incident channel: %s", err.Error()), false)); sendErr != nil {
			log.Error().Err(sendErr).Msg(sendError)
			return
		}
	}
	if _, err := h.slackClient.SetTopicOfConversationContext(ctx, incidentChannel.ID, overview); err != nil {
		if sendErr := h.sendMessage(ctx, params.broadcastChannel, slack.MsgOptionPostEphemeral(params.incidentDeclarer),
			slack.MsgOptionText(fmt.Sprintf("Failed to set topic for incident channel: %s", err.Error()), false)); sendErr != nil {
			log.Error().Err(sendErr).Msg(sendError)
			return
		}
	}
	// Add invitees to channel - the InviteUsersToConversationContext method
	// does not accept group as user so have to specify users individually
	if _, err := h.slackClient.InviteUsersToConversationContext(ctx, incidentChannel.ID, params.incidentInvitees...); err != nil {
		if err.Error() != alreadyInChannel {
			if sendErr := h.sendMessage(ctx, params.broadcastChannel, slack.MsgOptionPostEphemeral(params.incidentDeclarer),
				slack.MsgOptionText(fmt.Sprintf("Failed to add invitees to incident channel: %s", err.Error()), false)); sendErr != nil {
				log.Error().Err(sendErr).Msg(sendError)
				return
			}
		}
	}
	// Inform about incident
	if err := h.sendMessage(ctx, params.broadcastChannel,
		slack.MsgOptionText(fmt.Sprintf(":rotating_siren: An incident has been declared by <@%s>\n"+
			"*Incident summary:* %s\n"+
			"*Environment affected:* %s\n"+
			"*Region affected:* %s\n"+
			"*Severity:* %s\n"+
			"*Impact:* %s\n"+
			"*Responder:* <@%s>\n"+
			"*Commander:* <@%s>\n"+
			"*Incident channel:* <#%s>\n"+
			securityMessage,
			params.incidentDeclarer, params.incidentSummary, params.incidentEnvironmentsAffected,
			params.incidentRegionsAffected, params.IncidentSeverityLevel, params.IncidentImpactLevel,
			params.incidentResponder, params.incidentCommander,
			incidentChannel.ID), false)); err != nil {
		log.Error().Err(err).Msg(sendError)
		return
	}
	// Send message about starting a video call for live troubleshooting
	if err := h.sendMessage(ctx, incidentChannel.ID,
		slack.MsgOptionText(fmt.Sprintf("IC <@%s>: Start an incident Teams call with the command `/teams-calls meeting %s` and invite the appropriate people", params.incidentCommander, params.incidentChannelName), false)); err != nil {
		log.Error().Err(err).Msg(sendError)
		return
	}
	// Send message about starting an incident document for postmortem
	if err := h.sendMessage(ctx, incidentChannel.ID,
		slack.MsgOptionText(fmt.Sprintf("IC <@%s>: Start the incident document by using <%s|this template>", params.incidentCommander, h.opts.IncidentDocTemplateURL), false)); err != nil {
		log.Error().Err(err).Msg(sendError)
		return
	}
	// Add channel reminder about updating progress
	// Need to use user access token since bot token is not allowed token type: https://api.slack.com/methods/reminders.add
	userSlackClient := slack.New(h.opts.UserAccessToken)
	user, err := h.slackClient.GetUserInfoContext(ctx, params.incidentDeclarer)
	if err != nil {
		if sendErr := h.sendMessage(ctx, incidentChannel.ID, slack.MsgOptionPostEphemeral(params.incidentDeclarer),
			slack.MsgOptionText(fmt.Sprintf("Failed to get user info context: %s", err.Error()), false)); sendErr != nil {
			log.Error().Err(sendErr).Msg(sendError)
			return
		}
	}
	loc := time.FixedZone("CUSTOM-TZ", user.TZOffset)
	now := time.Now().In(loc)
	if _, err := userSlackClient.AddChannelReminder(incidentChannel.ID,
		fmt.Sprintf("\"Reminder for IC <@%s>: Update progress about the incident every 30 min in <#%s>, or remove the reminder and archive the channel if the incident is resolved\"", params.incidentCommander, params.broadcastChannel),
		fmt.Sprintf("every day at %s", now.Add(time.Minute*time.Duration(30)).Format("03:04:05PM"))); err != nil {
		if sendErr := h.sendMessage(ctx, incidentChannel.ID, slack.MsgOptionPostEphemeral(params.incidentDeclarer),
			slack.MsgOptionText(fmt.Sprintf("Failed to add channel reminder: %s", err.Error()), false)); sendErr != nil {
			log.Error().Err(sendErr).Msg(sendError)
			return
		}
	}
}

type resolveParams struct {
	// Broadcast channel to announce resolvement
	broadcastChannel string
	// Incident channel ID
	incidentChannel string
	// Incident resolution
	incidentResolution string
	// Should the incident channel be archived
	incidentArchive bool
	// Whow is resolving the incident
	incidentResolver string
}

// resolveIncident - handler for resolving incidents
func (h *botHandler) resolveIncident(ctx context.Context, payload *slack.InteractionCallback, w http.ResponseWriter) error {
	incidentChannelID := payload.View.State.Values["incident_channel"]["incident_channel"].SelectedConversation
	incChannel, _ := h.slackClient.GetConversationInfoContext(ctx, incidentChannelID, false)
	if err := validateChosenIncidentChannelName("incident_channel", incChannel.Name); err != nil {
		var verr *validationError
		if errors.As(err, &verr) {
			return postErrorResponse(ctx, verr.errors, w)
		}
		return err
	}
	resolveParams := &resolveParams{
		broadcastChannel:   payload.View.State.Values["broadcast_channel"]["broadcast_channel"].SelectedOption.Value,
		incidentChannel:    incidentChannelID,
		incidentResolution: payload.View.State.Values["resolution"]["resolution"].Value,
		incidentArchive:    payload.View.State.Values["archive_choice"]["archive_choice"].SelectedOption.Value == "Yes",
		incidentResolver:   payload.User.ID,
	}

	w.WriteHeader(http.StatusAccepted)

	// Do the rest via goroutine
	ctx = wrappedcontext.WrapContextValues(context.Background(), ctx)
	go h.doResolveTasks(ctx, resolveParams)

	return nil
}

func (h *botHandler) doResolveTasks(ctx context.Context, params *resolveParams) {
	log := zerolog.Ctx(ctx)
	// Inform about resolution
	if err := h.sendMessage(ctx, params.broadcastChannel,
		slack.MsgOptionText(fmt.Sprintf(":white_check_mark: The incident <#%s> has been resolved!\n"+
			"*Resolution:* %s",
			params.incidentChannel, params.incidentResolution), false)); err != nil {
		log.Error().Err(err).Msg("Could not send failure message")
		return
	}
	if params.incidentArchive {
		if err := h.slackClient.ArchiveConversationContext(ctx, params.incidentChannel); err != nil {
			log.Error().Err(err).Msg("Could not archive channel")
			return
		}
	}
}

// sendMessage - a simplified way to send a message
func (h *botHandler) sendMessage(ctx context.Context, channelID string, options ...slack.MsgOption) error {
	log := zerolog.Ctx(ctx)
	ch, ts, txt, err := h.slackClient.SendMessageContext(ctx, channelID, options...)
	if err != nil {
		err = fmt.Errorf("failed to send message: %w", err)
		log.Error().Err(err).Str("channel", ch).Str("message_timestamp", ts).Str("message_text", txt).Send()
		return err
	}
	return nil
}

func postErrorResponse(ctx context.Context, verr map[string]string, w http.ResponseWriter) error {
	log := zerolog.Ctx(ctx)
	errorResponse := slack.NewErrorsViewSubmissionResponse(verr)
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	enc := json.NewEncoder(w)
	err := enc.Encode(errorResponse)
	if err != nil {
		log.Err(err).Send()
		return err
	}
	errorResponseData, err := json.Marshal(errorResponse)
	if err != nil {
		log.Err(err).Send()
		return err
	}
	log.Warn().Str("errorResponseData", string(errorResponseData)).Msg("Modal validation error")
	return nil
}

// createChannelName - create incident channel name based on input field
func createChannelName(s string) string {
	return fmt.Sprintf("inc_%s_%s", s, strings.ToLower(time.Now().Format("2Jan2006")))
}

// createUserFriendlyConversationError - Map https://api.slack.com/methods/conversations.create error codes to user friendly messages
func createUserFriendlyConversationError(err error) error {
	if err.Error() == "name_taken" {
		return fmt.Errorf("this channel already exists")
	}
	return err
}

// validateIncidentChannelName - validate channel name according to slack rules in https://api.slack.com/methods/conversations.create
func validateIncidentChannelName(field string, n string) error {
	errorMessage := make(map[string]string)
	if !channelNameRegex.MatchString(n) {
		errorMessage[field] = fmt.Sprintf(valIncChName, n)
		return &validationError{
			errors: errorMessage,
		}
	}
	return nil
}

func validateChosenIncidentChannelName(field string, n string) error {
	errorMessage := make(map[string]string)
	if !incChannelNameRegex.MatchString(n) {
		errorMessage[field] = fmt.Sprintf(valChosenIncChName, n)
		return &validationError{
			errors: errorMessage,
		}
	}
	return nil
}
